### PR TITLE
Fix os.urandom to return uniform first byte

### PR DESCRIPTION
**Pull request by zolstein**
The former implementation of os.urandom generated an N-byte random
BigInteger and got its contents as a byte array. However, because the
byte array returned by a BigInteger can have one additional byte to
account for the sign, it truncated the array to only the first N bytes.

This causes a problem with the first byte; any time the high bit of
the first byte would have been a 1, it would instead make the entire
first byte 0.

I've rewritten the urandom function not to use a BigInteger and instead
use the Random.nextBytes method. This both fixes the bug and likely
makes the code faster because it removes the array-copy needed to
truncate the array.

### CODE REVIEW

**Comment by zolstein** [comment]
A coworker pointed out to me that Java's `Random` object may not be sufficient to implement `os.urandom`, which should return numbers random enough to use for cryptographic purposes. Possibly this code should use `SecureRandom` rather than `Random`. I don't actually have enough familiarity with `SecureRandom` to know whether there's a good reason _not_ to use it here, so I won't change this commit yet, but you should consider that and change it (or tell me to change it) if it's the right thing to do.

**Comment by timfel** [comment]
Hello, thanks for the PR. We will get to it shortly.
Regarding urandom vs random, there's a good (and still relevant) article here: https://www.2uo.de/myths-about-urandom/
TL;DR: urandom and random are pretty much the same for anything that runs more than few seconds after the system started

**Comment by zolstein** [comment]
I think I ran across that article while looking into this bug. My concern isn't with random vs urandom but with Java's Random class. The [python docs on os.urandom](https://docs.python.org/3/library/os.html#os.urandom) describe that:

>This function returns random bytes from an OS-specific randomness source. The returned data should be unpredictable enough for cryptographic applications, though its exact quality depends on the OS implementation.
>On Linux, if the getrandom() syscall is available, it is used in blocking mode: block until the system urandom entropy pool is initialized (128 bits of entropy are collected by the kernel). See the PEP 524 for the rationale. On Linux, the getrandom() function can be used to get random bytes in non-blocking mode (using the GRND_NONBLOCK flag) or to poll until the system urandom entropy pool is initialized.
>On a Unix-like system, random bytes are read from the /dev/urandom device. If the /dev/urandom device is not available or not readable, the NotImplementedError exception is raised.

This seems to match with Java's description of how [SecureRandom](https://docs.oracle.com/javase/8/docs/api/java/security/SecureRandom.html) generates its seed.

>Depending on the implementation, the generateSeed and nextBytes methods may block as entropy is being gathered, for example, if they need to read from /dev/random on various Unix-like operating systems.

By comparison, here's the complete implementation of how [Random](http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/tip/src/share/classes/java/util/Random.java) generates its seed value:
```java
    /**
     * Creates a new random number generator. This constructor sets
     * the seed of the random number generator to a value very likely
     * to be distinct from any other invocation of this constructor.
     */
    public Random() {
        this(seedUniquifier() ^ System.nanoTime());
    }

    private static long seedUniquifier() {
        // L'Ecuyer, "Tables of Linear Congruential Generators of
        // Different Sizes and Good Lattice Structure", 1999
        for (;;) {
            long current = seedUniquifier.get();
            long next = current * 181783497276652981L;
            if (seedUniquifier.compareAndSet(current, next))
                return next;
        }
    }

    private static final AtomicLong seedUniquifier
        = new AtomicLong(8682522807148012L);
```

I'm definitely not a cryptography expert, but the Random implementation looks substantially weaker than what the Python and SecureRandom docs describe.

**Comment by nirvdrum** [comment]
I agree with @zolstein. `java.util.Random` is a PRNG and doesn't read from their /dev/random or /dev/urandom. As for which source to use, I'd expect most callers of `urandom` would expect `/dev/urandom` in order to avoid blocking calls. `SecureRandom` can be [configured to source from /dev/urandom](https://github.com/oracle/truffleruby/blob/122eeb36b159b5938d6212fa58b0592f60cf4253/src/main/java/org/truffleruby/RubyContext.java#L743-L751), which is what we do in TruffleRuby. Then you can just [request the number of bytes you need](https://github.com/oracle/truffleruby/blob/122eeb36b159b5938d6212fa58b0592f60cf4253/src/main/java/org/truffleruby/RubyContext.java#L753-L758). Just to round the thought ought, we go a step further and wrap it up as a [VM primitive](https://github.com/oracle/truffleruby/blob/122eeb36b159b5938d6212fa58b0592f60cf4253/src/main/java/org/truffleruby/core/VMPrimitiveNodes.java#L398-L418),

**Comment by msimacek** [comment]
We got similar feedback from our security engineers. @zolstein, please, use `SecureRandom` like in @nirvdrum's example.

**Comment by msimacek** [comment]
Thank you for the contribution
